### ISSUE TITLE
Templating, `dart_model` and augmentation language version

### ISSUE THREAD

**Issue by davidmorgan** [dart-lang/macros#70]
In [this PR](https://github.com/dart-lang/macros/pull/68/files#diff-a41147bed1d6683f0d64dac93fa9c97cc2399c0d3e176b9db7ff05dfb9765d40) I use `{{uri#name}}` in macro-generated augmentation "code" to signal types that need to be resolved, as suggested by Remi [here](https://github.com/dart-lang/language/issues/1989#issuecomment-2329438841).

This does make the macro code look a lot nicer and I guess we'll want something along these lines if it's workable.

Concerns :)

1. Performance. I'm guessing it's trivially fast enough, but it's worth checking.
1. Representation on the wire, versioning.

By "representation on the wire" I mean: are templates parsed on the client then sent as structured data to the host, or are they sent as strings to the host and parsed there? Then, crucially: what happens if the template format changes?

Now that I think about this, I actually think it's the same problem as language versioning. It can be the case that an augmentation is valid Dart 3 code but not valid Dart 4 code. So, sending code as raw strings is already a problem: what we should be doing is tagging the augmentation with a language version. So the macro can say "this is Dart 3.3 code" and the host, if it needs Dart 4 code, knows it must either convert 3.3->4 or fail.

And if we're doing that we can at the same time specify a template language. So the macro can say "this is Dart 3.3 code before v1 template expansion". And then the host knows how to deal with that.

**Comment by jakemac53** [comment]
I don't think we want a templating thing like this baked into the macro support - it can happily live on the client side in its own library (as Remi did). We can probably support resolving many of these in a single query so it will also be fast enough.

Longer term, this type of template should go away entirely, and be replaced by user created identifiers + tagged strings.
